frame/assets/src/functions.rs

```diff
@@ -652,71 +661,125 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
-				debug_assert_eq!(details.sufficients, 0);
+				remaining_accounts = details.accounts;
+				Ok(())
+			})?;
+
+		for who in dead_accounts {
+			T::Freezer::died(id, &who);
+		}
+
+		Self::deposit_event(Event::DestroyedAccounts {
+			asset_id: id,
+			accounts_destroyed: removed_accounts as u32,
+			accounts_remaining: remaining_accounts as u32,
+		});
+		Ok(removed_accounts)
+	}
+
+	/// Destroy approvals associated with a given asset up to the max (T::RemoveItemsLimit).
+	///
+	/// Each call emits the `Event::DestroyedApprovals` event
+	pub(super) fn do_destroy_approvals(id: T::AssetId) -> Result<u32, DispatchError> {
+		let mut removed_approvals = 0;
+		let _ =
+			Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
+				let mut details = maybe_details.as_mut().ok_or(Error::<T, I>::Unknown)?;
+
+				ensure!(details.is_frozen, Error::<T, I>::NotFrozen);
+				// Should only destroy accounts while the asset is being destroyed
+				ensure!(details.status == AssetStatus::Destroying, Error::<T, I>::LiveAsset);
+
+				for ((owner, _), approval) in Approvals::<T, I>::drain_prefix((id,)) {
+					T::Currency::unreserve(&owner, approval.deposit);
+					removed_approvals = removed_approvals.saturating_add(1);
+					details.approvals = details.approvals.saturating_sub(1);
+					if removed_approvals >= T::RemoveItemsLimit::get() {
+						break
+					}
+				}
+				Self::deposit_event(Event::DestroyedApprovals {
+					asset_id: id,
+					approvals_destroyed: removed_approvals as u32,
+					approvals_remaining: details.approvals as u32,
+				});
+				Ok(())
+			})?;
+		Ok(removed_approvals)
+	}
+
+	/// Complete destrouing asset and unreserve the currency.
+	///
+	/// On success, the `Event::Destroyed` event is emitted.
+	pub(super) fn do_finish_destroy(id: T::AssetId) -> DispatchResult {
+		let _ =
+			Asset::<T, I>::try_mutate_exists(id, |maybe_details| -> Result<(), DispatchError> {
+				let details = maybe_details.take().ok_or(Error::<T, I>::Unknown)?;
+				ensure!(details.is_frozen, Error::<T, I>::NotFrozen);
+				ensure!(details.status == AssetStatus::Destroying, Error::<T, I>::LiveAsset);
+				ensure!(details.accounts == 0, Error::<T, I>::InUse);
+				ensure!(details.approvals == 0, Error::<T, I>::InUse);
 
 				let metadata = Metadata::<T, I>::take(&id);
 				T::Currency::unreserve(
 					&details.owner,
 					details.deposit.saturating_add(metadata.deposit),
 				);
-
-				for ((owner, _), approval) in Approvals::<T, I>::drain_prefix((&id,)) {
-					T::Currency::unreserve(&owner, approval.deposit);
-				}
 				Self::deposit_event(Event::Destroyed { asset_id: id });
 
-				Ok(DestroyWitness {
-					accounts: details.accounts,
-					sufficients: details.sufficients,
-					approvals: details.approvals,
-				})
-			},
-		)?;
-
-		// Execute hooks outside of `mutate`.
-		for who in dead_accounts {
-			T::Freezer::died(id, &who);
-		}
-		Ok(result_witness)
+				Ok(())
+			})?;
+		Ok(())
```

Comment:
```suggestion
			})
```
And without the `let _ =`.